src/messages/MOSDOp.h

```diff
@@ -364,9 +364,38 @@ struct ceph_osd_request_head {
 
       encode(retry_attempt, payload);
       encode(features, payload);
-    } else {
-      // latest v8 encoding with hobject_t hash separate from pgid, no
+    } else if (!HAVE_FEATURE(features, SERVER_SQUID)) {
+      // v8 encoding with hobject_t hash separate from pgid, no
       // reassert version
+      header.version = 8;
+
+      encode(pgid, payload);
+      encode(hobj.get_hash(), payload);
+      encode(osdmap_epoch, payload);
+      encode(flags, payload);
+      encode(reqid, payload);
+      encode_trace(payload, features);
```

Comment:
I would love to not have the two-staged msg decode, and thus be able to put this at the very end.

Reply:
yes, this would simplify the compatibility issues significantly... but I don't think it should be discussed here :-)

Reply:
Agreed, it's a separated discussion. I will ping @markhpc in the Perf Weekly. The numbers we have on the optimization pretty ancient.